### ISSUE TITLE
Failure to Precompile: JSString not defined

### ISSUE THREAD

**Issue by bnels** [sglyon/PlotlyJS.jl#198]
After updating my packages this morning, when precompiling PlotlyJS, I ran into the following issue:

```
using PlotlyJS
INFO: Precompiling module PlotlyJS.
ERROR: LoadError: LoadError: LoadError: UndefVarError: JSString not defined
Stacktrace:
 [1] include_from_node1(::String) at ./loading.jl:576
 [2] include(::String) at ./sysimg.jl:14
 [3] include_from_node1(::String) at ./loading.jl:576
 [4] include(::String) at ./sysimg.jl:14
 [5] include_from_node1(::String) at ./loading.jl:576
 [6] include(::String) at ./sysimg.jl:14
 [7] anonymous at ./<missing>:2
while loading /home/brad/.julia/v0.6/PlotlyJS/src/displays/electron.jl, in expression starting on line 228
while loading /home/brad/.julia/v0.6/PlotlyJS/src/display.jl, in expression starting on line 124
while loading /home/brad/.julia/v0.6/PlotlyJS/src/PlotlyJS.jl, in expression starting on line 34
ERROR: Failed to precompile PlotlyJS to /home/brad/.julia/lib/v0.6/PlotlyJS.ji.
Stacktrace:
 [1] compilecache(::String) at ./loading.jl:710
 [2] _require(::Symbol) at ./loading.jl:497
 [3] require(::Symbol) at ./loading.jl:405
```

Here is my `versioninfo()`
```
Julia Version 0.6.4
Commit 9d11f62bcb (2018-07-09 19:09 UTC)
Platform Info:
  OS: Linux (x86_64-redhat-linux)
  CPU: Intel(R) Core(TM) i7-6820HQ CPU @ 2.70GHz
  WORD_SIZE: 64
  BLAS: libopenblas (USE64BITINT DYNAMIC_ARCH NO_AFFINITY Haswell MAX_THREADS=16)
  LAPACK: libopenblas64_
  LIBM: libopenlibm
  LLVM: libLLVM-3.9.1 (ORCJIT, skylake)
```

**Comment by haotian127** [comment]
same issue +1

**Comment by judreg** [comment]
I have the same problem, too.

**Comment by Eetion** [comment]
same as well.

**Comment by irhooper** [comment]
And another one.

**Comment by sglyon** [comment]
I have this fixed locally, but some other things are broken given recent changes on the Blink/WebIO side. 

I think it is about time to merge in the display code from https://github.com/sglyon/PlotlyWebIO.jl. This will resolve all the issues here. I just need to find the time to do it.

If anyone is willing to contribute, the roadmap would be 

- replace the current SyncPlot implementation with the `WebIOPlot` type (probably keeping the name `SyncPlot`)
- Replicate the current functionality from the display files (`display.jl` and all the ones that it includes)

**Comment by sglyon** [comment]
Two more comments:

- The most recent commit on master (bbd16953913a7842e9009167273d3b7c548d91be) fixes this particular problem, but some other things are still broken 
- If anyone does want to make an attempt at integrating the code from PlotlyWebIO please leave a message here so we can coordinate efforts.

Th

**Comment by T-Doc** [comment]
Tested the commit, and it compiles, but it won't let me display plots. Creating the plot object works, but it fails when trying to display it:

>                _
>    _       _ _(_)_     |  A fresh approach to technical computing
>   (_)     | (_) (_)    |  Documentation: https://docs.julialang.org
>    _ _   _| |_  __ _   |  Type "?help" for help.
>   | | | | | | |/ _` |  |
>   | | |_| | | | (_| |  |  Version 0.6.4 (2018-07-09 19:09 UTC)
>  _/ |\__'_|_|_|\__'_|  |  Official http://julialang.org/ release
> |__/                   |  x86_64-pc-linux-gnu
> 
> julia> using PlotlyJS
> INFO: Loading HttpServer methods...
> 
> julia> p1 = Plot(scatter(;y=randn(3)))
> data: [
>   "scatter with fields type and y"
> ]
> 
> layout: "layout with field margin"
> 
> 
> 
> julia> plot(p1)
> Error showing value of type PlotlyJS.SyncPlot{PlotlyJS.ElectronDisplay}:
> ERROR: JSExpr: Unsupported `error` expression, $(Expr(:error, UndefVarError(Symbol("@new"))))
> Stacktrace:
>  [1] macro expansion at /home/tbreslein/.julia/v0.6/JSExpr/src/JSExpr.jl:191 [inlined]
>  [2] macro expansion at /home/tbreslein/.julia/v0.6/MacroTools/src/macro.jl:18 [inlined]
>  [3] jsexpr(::Expr) at /home/tbreslein/.julia/v0.6/JSExpr/src/JSExpr.jl:159
>  [4] macro expansion at /home/tbreslein/.julia/v0.6/JSExpr/src/JSExpr.jl:179 [inlined]
>  [5] macro expansion at /home/tbreslein/.julia/v0.6/MacroTools/src/macro.jl:18 [inlined]
>  [6] jsexpr(::Expr) at /home/tbreslein/.julia/v0.6/JSExpr/src/JSExpr.jl:159
>  [7] jsstring at /home/tbreslein/.julia/v0.6/JSExpr/src/JSExpr.jl:31 [inlined]
>  [8] #js#4(::Bool, ::Function, ::Blink.AtomShell.Window, ::Expr) at /home/tbreslein/.julia/v0.6/Blink/src/rpc/rpc.jl:45
>  [9] js(::Blink.AtomShell.Window, ::Expr) at /home/tbreslein/.julia/v0.6/Blink/src/rpc/rpc.jl:45
>  [10] #get_window#46(::Array{Any,1}, ::Function, ::PlotlyJS.ElectronDisplay) at /home/tbreslein/.julia/v0.6/PlotlyJS/src/displays/electron.jl:68
>  [11] (::PlotlyJS.#kw##get_window)(::Array{Any,1}, ::PlotlyJS.#get_window, ::PlotlyJS.ElectronDisplay) at ./<missing>:0
>  [12] #get_window#47(::Array{Any,1}, ::Function, ::PlotlyJS.SyncPlot{PlotlyJS.ElectronDisplay}) at /home/tbreslein/.julia/v0.6/PlotlyJS/src/displays/electron.jl:109
>  [13] (::PlotlyJS.#kw##get_window)(::Array{Any,1}, ::PlotlyJS.#get_window, ::PlotlyJS.SyncPlot{PlotlyJS.ElectronDisplay}) at ./<missing>:0
>  [14] #display_blink#49(::Bool, ::Bool, ::Function, ::PlotlyJS.SyncPlot{PlotlyJS.ElectronDisplay}) at /home/tbreslein/.julia/v0.6/PlotlyJS/src/displays/electron.jl:116
>  [15] (::PlotlyJS.#kw##display_blink)(::Array{Any,1}, ::PlotlyJS.#display_blink, ::PlotlyJS.SyncPlot{PlotlyJS.ElectronDisplay}) at ./<missing>:0
>  [16] display(::PlotlyJS.SyncPlot{PlotlyJS.ElectronDisplay}) at /home/tbreslein/.julia/v0.6/PlotlyJS/src/displays/electron.jl:112
>  [17] eval(::Module, ::Any) at ./boot.jl:235
>  [18] print_response(::Base.Terminals.TTYTerminal, ::Any, ::Void, ::Bool, ::Bool, ::Void) at ./REPL.jl:144
>  [19] print_response(::Base.REPL.LineEditREPL, ::Any, ::Void, ::Bool, ::Bool) at ./REPL.jl:129
>  [20] (::Base.REPL.#do_respond#16{Bool,Base.REPL.##26#36{Base.REPL.LineEditREPL,Base.REPL.REPLHistoryProvider},Base.REPL.LineEditREPL,Base.LineEdit.Prompt})(::Base.LineEdit.MIState, ::Base.AbstractIOBuffer{Array{UInt8,1}}, ::Bool) at ./REPL.jl:646
> 


Same when using it through Plots, though that should hardly be a surprise.

**Comment by aaronsheldon** [comment]
How much of this can be accomplished by cutting out all the other displays so that only the WebIO display is used and then replacing all the HTML/Script returning functions with functions that return WebIO.nodes? Or maybe I'm missing something?

**Comment by sglyon** [comment]
@aaronsheldon that is what we need to do. 

The comment above about integrating PlotlyWebIO.jl is exactly that.

It has code only against WebIO for display on all front ends. Then we just need to connect other bells/whistles to WebIO (much of it is already done) to implement all the features currently in this package.

I simply don’t have the cycles right now to do the integration. If someone else wants to open a pull request that does it I would be happy to review/refine/merge.

**Comment by aaronsheldon** [comment]
Okay. I only have evenings right now, but I really want this package to work. I use it quite a bit. I'll scratch away at, and hopefully anyone more skilled than me will beat me to the punch.

**Comment by sglyon** [comment]
Ok thanks for understanding and expressing your interest!

One some of the Julia 0.7/1.0 dust settles I imagine that packages will stabailze and this should be a pretty easy transition.

**Comment by aaronsheldon** [comment]
hmmm...WebIO is looking even more intriguing. I think data, layout, etc... could be communicated directly without JSString coercion using the scope/observable idiom of WebIO. So that the pipeline would roughly be:

1. Reference Plotly library in the WebIO scope
2. Setup observables for communicating data, layout, etc... variables
3. Setup JS side listeners to update JS side data, layout, etc... variables with changes in the Julia data, layout, etc...
4. Build a WebIO node DIV for the plot
5. Populate the Julia side data, layout, etc... variables
6. Call Plotly.plot on the JS side

I'm wondering if I'm misinterpreting WebIO, or making this overly complicated?

**Comment by lewisl** [comment]
Me, too.

so, PlotlyJS is dead for Julia 0.6.x and Julia 0.7 and Julia 1.0.

I can't even revert.

is there some way to go back to old versions?

**Comment by laborg** [comment]
Two weeks ago I had luck with this configuration:
https://discourse.julialang.org/t/how-are-people-working-around-the-current-plotlyjs-jl-outage/12835/2?u=laborg
Not sure if this still works.

**Comment by standarddeviant** [comment]
@laborg I appreciate the info/cross-link!

That 'recipe' seems more comprehensive than what I did:
```julia
Pkg.pin("Blink", v"0.6.2")
Pkg.pin("PlotlyJS", v"0.10.2")
Pkg.build("Blink")
```

As of now, I'm up and running with the `Plots.jl` interface of `PlotlyJS.jl` on 0.6.4 on Windows10, x86-64.  I'm using this via jupyter-lab via Anaconda.

**Comment by sglyon** [comment]
All, thank you for the reports and discussion.

I'm sorry this has been slow moving, but we are nearly there. There has been a great community effort to get the Web stack ready to go. Once that is in place we can make a few modifications here and all will be working again.

Please bear with us!

**Comment by sglyon** [comment]
See #208 for the latest
--

**Comment by sglyon** [comment]
For now, please use the master branch

```julia
] # enter pkg mode
(v1.0) pkg> checkout PlotlyJS
```

If you want save figures you can do

```
] # enter package mode
(v1.0) pkg> add https://github.com/sglyon/ORCA.jl.git#master
```

If this doesn't resolve your issue here, please reopen the issue and leave a comment describing what went wrong
